src/libopensc/cards.h

```diff
@@ -244,6 +244,7 @@ enum {
 	SC_CARD_TYPE_IDPRIME_3810,
 	SC_CARD_TYPE_IDPRIME_830,
 	SC_CARD_TYPE_IDPRIME_930,
+	SC_CARD_TYPE_IDPRIME_930_2,
```

Comment:
If I read it right, the new ATR matches basically the IDPrime 930, but it does support EC?

In that case, I would propose to use some better name to differentiate from the "short" IDPrime ATRs. Your ATR almost matches the `SC_CARD_TYPE_IDPRIME_GENERIC`, which was added as part of #2812, where the token is referenced as `eToken 5110`. Is there a way to make it more generic ATR mask to match both of them? Tagging @jurajsarinay to have a look or get some clarification if his eToken supports EC or his ATR can be used for some simplification here.

Reply:
First I tried to match SC_CARD_TYPE_IDPRIME_940 atr, but there are couple cases where it breaks.
This device does not have  0005 file
https://github.com/OpenSC/OpenSC/pull/3048/files#diff-1892ee089e440f54f73abfe799748d7cdbdf540419ae7898086382caeb395b43L582

And also I am not sure the key_reference arithmetic? how I can verify my device correct key_reference
https://github.com/OpenSC/OpenSC/pull/3048/files#diff-1892ee089e440f54f73abfe799748d7cdbdf540419ae7898086382caeb395b43L455
Maybe I can use the 940 key_reference case here
https://github.com/OpenSC/OpenSC/pull/3048/files#diff-1892ee089e440f54f73abfe799748d7cdbdf540419ae7898086382caeb395b43L457

Reply:
```
Token category: Hardware
Reader name: SafeNet eToken 5110+ FIPS 0
Serial number (PKCS#11): ....
Free space: 53788
Card ID (GUID): 0x...
Product name: eToken 5110+ FIPS
Card type: IDPrime
Applet Version: IDPrime Java Applet 4.5.0.F
Mask version: G286
Token Password: Present
Token Password retries remaining: 5
Maximum Token Password retries: 5
Token Password expiration: No expiration
Administrator Password: Present
Administrator Password retries remaining: 5
Maximum administrator Password retries: 5
FIPS: FIPS 140-2 L2
Full Secure Messaging (SM): No
Sign padding on-board: Yes
Supported key size: 4096 bits
ECC: Supported
CSP: eToken Base Cryptographic Provider
KSP: SafeNet Smart Card Key Storage Provider
```

Reply:
> And also I am not sure the key_reference arithmetic? how I can verify my device correct key_reference

Usually by running the `pkcs11-tool --test` or `p11test` with the token and having more than one keys on the token loaded. The tools do signature+verification process  and if the key identifier would not match the calculation, the signature would either fail (if wrong key reference would be used) or the signature verification would fail (if the key reference matching different key than we think would be used).

Reply:
Seems like this test is RSA only
```
pkcs11-tool --test
Using slot 0 with a present token (0x0)
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  seems to be OK
Digests:
  all 4 digest functions seem to work
Failed to load legacy provider
Ciphers: not implemented
Signatures (currently only for RSA)
Signatures: no private key found in this slot
Verify (currently only for RSA)
  No private key found for testing
Unwrap: not implemented
Decryption (currently only for RSA)
No errors
```

Reply:
Yes, pkcs11-tool test supports only RSA. If you want more extensive tests, I can recommend the p11test in `src/tests/p11test` (not packaged, needs to be built from source).